### PR TITLE
SSCS-10332 - Ability to book hearings for 'paper' cases to any venue within the same region as the RPC

### PR DESCRIPTION
**Pull request by DGordonV1**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SSCS-10332

### Change description ###

For paper cases, retrieve list of epimsIds with the same RPC name of existing CCD RPC, returning a list of HearingLocation objects populated with these epimsIds.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java [hunk 1]
```diff
@@ -201,19 +201,32 @@ public static List<HearingLocation> getHearingLocations(SscsCaseData caseData,
...
-            .findFirst()
-            .orElseGet(() -> Collections.singletonList(epimsId))
-            .stream().map(epims -> HearingLocation.builder().locationId(epims).locationType(COURT).build())
-            .collect(Collectors.toCollection(ArrayList::new));
+        if (HearingsChannelMapping.isPaperCase(caseData)) {
+            List<String> epimsIds = referenceDataServiceHolder
+                .getVenueService()
+                .getActiveRegionalEpimsIdsForRpc(caseData.getRegionalProcessingCenter().getName());
+
+            return epimsIds.stream()
```
Comment (dattipoeHMCTS) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java:209]: (Minor) could perhaps merge the final stream of collect in both paths to remove duplication

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 2]
```diff
@@ -104,6 +83,7 @@ void buildHearingDetails() {
                                         .baseLocation(EPIMS_ID)
                                         .region(REGION)
                                         .build())
+            .dwpIsOfficerAttending("Yes")
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:86]: Where did these two changes come from?

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 3]
```diff
@@ -272,6 +254,36 @@ void getMultipleHearingLocations_shouldReturnCorrespondingMultipleEpimsIdForVenu
...
+    void getRegionalHearingLocations_shouldReturnCorrespondingEpimsIdsForVenuesWithSameRpc() {
+        final SscsCaseData caseData = SscsCaseData.builder()
+            .dwpIsOfficerAttending("No")
+            .regionalProcessingCenter(RegionalProcessingCenter.builder().name("SSCS Leeds").build())
+            .appeal(Appeal.builder()
+                        .hearingOptions(HearingOptions.builder().wantsToAttend("N").build())
+                        .build())
+            .processingVenue(PROCESSING_VENUE_1)
+            .build();
+        given(venueService.getActiveRegionalEpimsIdsForRpc(caseData.getRegionalProcessingCenter().getName())).willReturn(Arrays.asList("112233","332211","123123","321321"));
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:268]: (Minor) line length

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 4]
```diff
@@ -272,6 +254,36 @@ void getMultipleHearingLocations_shouldReturnCorrespondingMultipleEpimsIdForVenu
...
+            caseData,
+            referenceDataServiceHolder
+        );
+
+        assertThat(result).hasSize(4);
+        assertThat(result)
+            .hasSize(4)
+            .extracting("locationId","locationType")
+            .containsExactlyInAnyOrder(
+                tuple("112233", COURT),
```
Comment (dattipoeHMCTS) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:281]: Could move the ids into variables

**Review by ss-evoco — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java (hunk 2 above)
Comment (DGordonV1) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:86]: Quick fix to some tests so they don't fall into the isPaperCase() path of new code.

**Review by dattipoeHMCTS — APPROVED** [review]

**Review by amittal89 — DISMISSED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java [hunk 5]
```diff
@@ -201,19 +202,33 @@ public static List<HearingLocation> getHearingLocations(SscsCaseData caseData,
...
-            .stream().map(epims -> HearingLocation.builder().locationId(epims).locationType(COURT).build())
-            .collect(Collectors.toCollection(ArrayList::new));
+            return venueDetailsList.stream()
+                .map(VenueDetails::getEpimsId)
+                .map(id -> HearingLocation.builder()
+                    .locationId(id)
+                    .locationType(COURT)
+                    .build())
+                .collect(Collectors.toList());
+        } else {
```
Comment (TeaWithLucas) [inline src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java:217]: You don't need an else here as you return in the if statement and removing will make the code cleaner

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 6]
```diff
@@ -8,30 +8,10 @@
...
-import uk.gov.hmcts.reform.sscs.ccd.domain.Name;
-import uk.gov.hmcts.reform.sscs.ccd.domain.OtherParty;
-import uk.gov.hmcts.reform.sscs.ccd.domain.OverrideFields;
-import uk.gov.hmcts.reform.sscs.ccd.domain.Party;
-import uk.gov.hmcts.reform.sscs.ccd.domain.Role;
-import uk.gov.hmcts.reform.sscs.ccd.domain.SchedulingAndListingFields;
-import uk.gov.hmcts.reform.sscs.ccd.domain.SessionCategory;
-import uk.gov.hmcts.reform.sscs.ccd.domain.SscsCaseData;
-import uk.gov.hmcts.reform.sscs.ccd.domain.YesNo;
+import uk.gov.hmcts.reform.sscs.ccd.domain.*;
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:11]: We don't use wildcard imports

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 7]
```diff
@@ -272,6 +265,37 @@ void getMultipleHearingLocations_shouldReturnCorrespondingMultipleEpimsIdForVenu
 
     }
 
+    @DisplayName("Regional hearing location Test")
+    @Test
+    void getRegionalHearingLocations_shouldReturnCorrespondingEpimsIdsForVenuesWithSameRpc() {
+        final SscsCaseData caseData = SscsCaseData.builder()
+            .dwpIsOfficerAttending("No")
+            .regionalProcessingCenter(RegionalProcessingCenter.builder().name("SSCS Leeds").build())
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:273]: Minor, can you put anything after builder() on a new line for readability and to make adding fields in the future easier

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 8]
```diff
@@ -272,6 +265,37 @@ void getMultipleHearingLocations_shouldReturnCorrespondingMultipleEpimsIdForVenu
 
     }
 
+    @DisplayName("Regional hearing location Test")
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:268]: Can you describe the test, what method is this unit test for? What are the inputs that cause what outcome?

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java [hunk 9]
```diff
@@ -77,6 +57,16 @@ class HearingsDetailsMappingTest extends HearingsMappingBase {
 
     public static final String PROCESSING_VENUE_1 = "test_place";
 
+    private static final String epimsId1 = "112233";
+    private static final String epimsId2 = "332211";
+    private static final String epimsId3 = "123123";
+    private static final String epimsId4 = "321321";
+    private static final List<VenueDetails> epimsIdsList = Arrays.asList(
```
Comment (TeaWithLucas) [inline src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java:64]: This should be set in a @BeforeEach setup method

**Review by TeaWithLucas — CHANGES_REQUESTED** [review]

**Review by TeaWithLucas — DISMISSED** [review]

**Review by dattipoeHMCTS — APPROVED** [review]

**Review by dattipoeHMCTS — DISMISSED** [review]

**Review by ikirsanov — APPROVED** [review]

**Review by TeaWithLucas — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java (hunk 1 above)
Comment (filippo-possenti-version1) [reply src/main/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMapping.java]: Done or not, we won't stop deployment on this.

File: src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java (hunk 4 above)
Comment (filippo-possenti-version1) [reply src/test/java/uk/gov/hmcts/reform/sscs/helper/mapping/HearingsDetailsMappingTest.java]: Agreed. We'll let it slide.

**Review by dattipoeHMCTS — APPROVED** [review]

**Review by TeaWithLucas — APPROVED** [review]

**Pull request merged**